src/android/com/microsoftopentechnologies/intellij/components/MSOpenTechToolsProject.java

```diff
@@ -0,0 +1,172 @@
+    private static final String TEMPLATE_ZIP_NAME = "templates.zip";
+    private static final String CACHED_TEMPLATE_ZIP_NAME = "templates-%s.zip";
+
+    protected MSOpenTechToolsProject(Project project) {
+        super(project);
+    }
+
+    private VirtualFileListener virtualFileListener = null;
+    private boolean templatesCreated = false;
+
+    @Override
+    public void projectOpened() {
+        try {
+            templatesCreated = false;
+
+            // if a new project is being created then this method is called before
+            // any file is created; we install a file listener to handle the template
+            // file creation in case the project being created is an Android gradle
+            // project
+            virtualFileListener = new VirtualFileAdapter() {
+                @Override
+                public void contentsChanged(VirtualFileEvent event) {
+                    super.fileCreated(event);
+
+                    // if this is a build.gradle file and is an android build file then
+                    // copy the template files
+                    try {
+                        if(!templatesCreated &&
+                                event.getFileName().contains("build.gradle") &&
+                                ServiceCodeReferenceHelper.isAndroidGradleBuildFile(event.getFile())) {
+                            createActivityTemplates();
+                        }
+                    } catch (IOException ignored) {}
+                }
+            };
+            VirtualFileManager.getInstance().addVirtualFileListener(virtualFileListener);
+
+            // get project root dir and check if this is an Android project
+            if (!ServiceCodeReferenceHelper.isAndroidGradleModule(myProject.getBaseDir())) {
+                return;
+            }
+
+            createActivityTemplates();
+
+        } catch (IOException ignored) {}
+    }
+
+    @Override
+    public void projectClosed() {
+        if(virtualFileListener != null) {
+            VirtualFileManager.getInstance().removeVirtualFileListener(virtualFileListener);
+        }
+    }
+
+    private void createActivityTemplates() throws IOException {
+        // create the root dir to contain our templates zip if the
+        // dir doesn't exist already
+        File rootDir = new File(myProject.getBaseDir().getPath(),
+                String.format(
+                        TEMPLATES_ROOT_DIR,
+                        MSOpenTechToolsApplication.getCurrent().getSettings().getPluginVersion()));
+        rootDir.mkdirs();
+
+        // we proceed only if "templates.zip" doesn't already exist in the path
+        File templatesZip = new File(rootDir, TEMPLATE_ZIP_NAME);
+        if(!templatesZip.exists()) {
+            File cachedZip = getTemplatesZip();
+            Files.copy(cachedZip, templatesZip);
+        }
+
+        templatesCreated = true;
+    }
+
```

Comment:
Yes the location and content seems to survive a clean and rebuild. This location is not in the app module's "build" folder - it is in root project folder's "build" folder. For some reason IntelliJ insists on looking in that particular location for activity templates. I didn't see any dependencies being added or anything being shipped in the apk. I am not entirely happy with the approach but it seems to be lesser of the two evils (the other being copying to the application's install location)!